### PR TITLE
Add support for torch.tensor_split to accept a tensor for `indices` argument

### PR DESCRIPTION
**Pull request by edqwerty10**
Summary:
Trying to solve PR request https://github.com/pytorch/pytorch/issues/47479.
This diff tries to overload method `torch.tensor_split` to also accept a tensor for argument `split_size_or_sections` which currently accepts a python list or int. The motivation is to avoid converting a tensor to a list so that when tracing a model/module the tensor operations can be recorded.

Implementation is following the diff that originally added the `tensor_split` method D24166164 (https://github.com/pytorch/pytorch/commit/ef4817fe5a16ba9969562911c5363736a1003bb0).

Test Plan:
```
buck test caffe2/test:others -- test_tensor_split
buck test caffe2/test:torch -- test_tensor_split_indices
 buck test caffe2/test:torch -- test_tensor_split_errors
```

Differential Revision: D25440885

### CODE REVIEW

**Comment by ngimel** [comment]
Test errors are real. Do you need help with those, @edqwerty10, or are you still working on the PR?

File: aten/src/ATen/native/TensorShape.cpp
Comment (ngimel) [inline aten/src/ATen/native/TensorShape.cpp:548]: please check that `indices_or_sections` is on cpu, it won't work otherwise (first branch will cause host-device sync, second branch will segfault when you are creating std::vector from tensor data)

Comment (edqwerty10) [reply aten/src/ATen/native/TensorShape.cpp:548]: Hi @ngimel , how would the code look like to check `indices_or_sections` is on cpu? Is this a TORCH_CHECK?
Once `indices_or_sections` is confirmed to be on cpu does the body of the code change?

I think we want to support this for gpu as well, how would the code change to support this?
cc. @xw285cornell

Comment (ngimel) [reply aten/src/ATen/native/TensorShape.cpp:548]: Yes, it should be a TORCH_CHECK, something like
```
TORCH_CHECK(tensor_indices_or_sections.device()==kCPU, "expected tensor_indices_or_sections to be on cpu")
```
After the check the body of the code does not need to change. Allowing `indices_or_sections` to be on gpu would require you to copy it to cpu here, which is a cross-device synchronization. Pytorch tries not to do implicit cross-device synchronization, if you need it, you can do it in the client code by calling `indices.cpu()`

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:551]: "expected tensor_indices_or_sections to be on cpu" -> "tensor_split() expected tensor_indices_or_sections to be on cpu, but it's on DEVICE_TYPE"

Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:551]: (where DEVICE_TYPE is populated with the current device of `tensor_indices_or_sections`)

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:557]: This check for kLong is good but should happen immediately before or after the check that `tensor_indices_or_sections` is on the CPU. Otherwise a scalar float tensor could be passed.

Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:557]: The check for dimensions can also happen along with the other checks. Check that the dimension is <= 1, and if it's not say it's expected to be zero or one dimensions and print how many dimensions it actually is, just like the comment for the device warning above.

File: test/test_view_ops.py
Comment (mruberry) [inline test/test_view_ops.py:1224]: This section of the tests will need to be updated to reflect the error checking changes mentioned above

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1354]: `indices_or_sections (int or (list(int))):` -> `indices_or_sections (Tensor, int or list or tuple of ints):`

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1355]: "... or scalar tensor ..." -> "... or a zero dimensional long tensor ..."

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1363]: "... or a one-dimensional tensor of ints ..." -> "... or a one-dimensional long tensor ..."

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1365]: "in the list" -> "in the list, tuple or tensor."

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1374]: These new examples are thorough but I think it's easy to infer what the behavior with passing a tensor will be. We probably don't need to change the examples.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:297]: I think this is the cause of the test failures. The splits must be on the CPU, but here they're being put on the passed device. Changing device to be "cpu" should work.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1367]: Add a new paragraph here that's just one sentence explaining the requirement that if this is a tensor it must be a CPU long tensor. Something like:

"If indices_or_sections is a tensor, it must be a zero-dimensional or one-dimensional long tensor on the CPU."

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:297]: There are a few other things to change with this SampleInput:

- add a second sample input where the indices_or_sections is a scalar tensor
- add a third sample input where the dim argument is used
- just create the indices_or_sections tensors directly instead of using `make_tensor`, that is, just call `torch.tensor(...` so the splits make sense given the input tensor. If the splits are randomly generated, for example, they might look very strange, like `(-5, 2, 1, ...)`

File: torch/testing/_internal/common_methods_invocations.py
Comment (ngimel) [reply torch/testing/_internal/common_methods_invocations.py:805]: #49356 should fix this, but it's ok to skip the test if we don't land it in time.

**Comment by mruberry** [comment]
Hey @edqwerty10, this is looking pretty good. I made some comments/suggestions inline, mostly about handling sample inputs and error checking. This also needs to be rebased because of changes in common_methods_invocations.py.

**Comment by mruberry** [comment]
The failures are real. 

- make_tensor now has to take an actual torch.device, which can be constructed like `torch.device('cpu')`. My mistake on suggesting just "cpu" would do.
- looks like bfloat16 is supported on CUDA, do the dtypesIfCUDA needs to be updated to include bfloat16

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:863]: #49356 has been merged. Try running the tests without this skip: it may no longer be needed.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1270]: Minus the quotation mark at the end of this sentence.

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:566]: This check should still be moved up or refined, because tensor_split expects `tensor_indices_or_sections` to be either a zero dimensional or one dimensional tensor. There's no user-facing concept of "indices" that the message can reference.

If the check is moved outside this if statement, for example, it would very that tensor_indices_or_sections has zero or 1 dimensions.

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:554]: Same type of formatting here where it starts with "tensor_split":

"tensor_split expected input to have at least one dimension, but got ..."

Comment (edqwerty10) [reply aten/src/ATen/native/TensorShape.cpp:554]: I kept it as is because the other two `tensor_split` signatures are raising the same exception message. If we change it, shall I change the other two signatures' messages as well?

Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:554]: OK, I appreciate that for consistency we can leave this as-is and not bother changing it.

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:518]: "got" -> "but got"

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:534]: "got" -> "but got"

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:549]: "got -> but got"

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:558]: "got -> but got"

File: test/test_view_ops.py
Comment (mruberry) [inline test/test_view_ops.py:1208]: "got" language will need to be updated here, too

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:555]: "got -> "but got"

The "but" is typical (like in the error message above) to emphasize the surprise at not getting what's expected.

File: test/test_view_ops.py
Comment (mruberry) [inline test/test_view_ops.py:1226]: These regexes will need updating per the above changes, too

**Review by mruberry — APPROVED** [review]
Nice work, @edqwerty10! I made a few final comments about some error message formatting that should be straightforward to implement.

In addition to those changes, let @ngimel take a look before merging this, too.

**Review by ngimel — APPROVED** [review]

**Comment by edqwerty10** [comment]
Thanks a lot for the help @mruberry  and @ngimel! :) I made the amends and I'm looking to land the PR/diff after the tests complete (amends will reflect on github soon). Do I have permission to do this? I see `Merging is blocked` which makes me think I do not. I can still hit the land button from Phabricator but looks like we need to merge from github side too, I think.

**Comment by mruberry** [comment]
> Thanks a lot for the help @mruberry and @ngimel! :) I made the amends and I'm looking to land the PR/diff after the tests complete (amends will reflect on github soon). Do I have permission to do this? I see `Merging is blocked` which makes me think I do not. I can still hit the land button from Phabricator but looks like we need to merge from github side too, I think.

You only need to "Ship It" from Phabricator once the tests finish and an automated system will mirror the changes to Github.

**Pull request closed without merging**
